src/cargo/ops/cargo_compile.rs

```diff
@@ -169,20 +178,44 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
         bail!("jobs must be at least 1")
     }
 
-    let profiles = root_package.manifest().profiles();
-    if spec.len() == 0 {
-        try!(generate_targets(root_package, profiles, mode, filter, release));
-    }
+    let spec: Cow<'a, [String]> = match spec {
+        Packages::Packages(spec) => spec.into(),
+        Packages::All => ws.members()
+                           .map(|package| {
+                               let package_id = package.package_id();
+                               PackageIdSpec::from_package_id(package_id).to_string()
+                            })
+                           .collect()
+    };
+
+    let profiles = match ws.current_opt() {
+        Some(package) => package.manifest().profiles().to_owned(),
+        None => Profiles {
+            release: Profile::default_release(),
+            dev: Profile::default_dev(),
+            test: Profile::default_test(),
+            test_deps: Profile::default_dev(),
+            bench: Profile::default_bench(),
+            bench_deps: Profile::default_release(),
+            doc: Profile::default_doc(),
+            custom_build: Profile::default_custom_build(),
+        },
+    };
```

Comment:
Hm this is somewhat worrisome to me. This is where I think we'll want to learn about the profiles from the workspace root crate instead of whatever the current crate happens to be.